src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -0,0 +1,731 @@
+    if (incomplete_->complete())
+    {
+        complete_.emplace(incomplete_->index(), std::move(incomplete_));
+        incomplete_.reset();
+        updateStats();
+    }
+    return true;
+}
+
+int
+DatabaseShardImp::getDesiredAsyncReadCount(std::uint32_t seq)
+{
+    auto const shardIndex {seqToShardIndex(seq)};
+    {
+        std::lock_guard<std::mutex> l(m_);
+        auto it = complete_.find(shardIndex);
+        if (it != complete_.end())
+            return it->second->pCache().getTargetSize() / asyncDivider;
+        if (incomplete_ && incomplete_->index() == shardIndex)
+            return incomplete_->pCache().getTargetSize() / asyncDivider;
+    }
+    return cacheTargetSize / asyncDivider;
+}
+
+float
+DatabaseShardImp::getCacheHitRate()
+{
+    float sz, f {0};
+    {
+        std::lock_guard<std::mutex> l(m_);
+        sz = complete_.size();
+        for (auto const& c : complete_)
+            f += c.second->pCache().getHitRate();
+        if (incomplete_)
+        {
+            f += incomplete_->pCache().getHitRate();
+            ++sz;
+        }
+    }
+    return f / std::max(1.0f, sz);
+}
+
+void
+DatabaseShardImp::tune(int size, int age)
+{
+    std::lock_guard<std::mutex> l(m_);
+    cacheSz_ = size;
+    cacheAge_ = age;
+    int const sz {calcTargetCacheSz()};
+    for (auto const& c : complete_)
+    {
+        c.second->pCache().setTargetSize(sz);
+        c.second->pCache().setTargetAge(cacheAge_);
+        c.second->nCache().setTargetSize(sz);
+        c.second->nCache().setTargetAge(cacheAge_);
+    }
+    if (incomplete_)
+    {
+        incomplete_->pCache().setTargetSize(sz);
+        incomplete_->pCache().setTargetAge(cacheAge_);
+        incomplete_->nCache().setTargetSize(sz);
+        incomplete_->nCache().setTargetAge(cacheAge_);
+    }
+}
+
+void
+DatabaseShardImp::sweep()
+{
+    std::lock_guard<std::mutex> l(m_);
+    int const sz {calcTargetCacheSz()};
+    for (auto const& c : complete_)
+    {
+        c.second->pCache().sweep();
+        c.second->nCache().sweep();
+        if (c.second->pCache().getTargetSize() > sz)
+            c.second->pCache().setTargetSize(sz);
+    }
+    if (incomplete_)
+    {
+        incomplete_->pCache().sweep();
+        incomplete_->nCache().sweep();
+        if (incomplete_->pCache().getTargetSize() > sz)
+            incomplete_->pCache().setTargetSize(sz);
+    }
+}
+
+std::shared_ptr<NodeObject>
+DatabaseShardImp::fetchFrom(uint256 const& hash, std::uint32_t seq)
+{
+    Backend* backend;
+    auto const shardIndex {seqToShardIndex(seq)};
+    {
+        std::unique_lock<std::mutex> l(m_);
+        auto it = complete_.find(shardIndex);
+        if (it != complete_.end())
+            backend = &it->second->getBackend();
+        else if (incomplete_ && incomplete_->index() == shardIndex)
+            backend = &incomplete_->getBackend();
+        else
+            return {};
+    }
+    return fetchInternal(hash, *backend);
+}
+
+// Lock must be held
```

Comment:
I don't see the lock acquired prior to calling.

Reply:
`findShardIndexToAdd` is called from `prepare`. The first line in `prepare` takes the lock.